storage/txnrecovery: add transaction recovery metrics

nvb opened the pull request:
This commit adds a series of time series metrics that provide insight
into the process of transaction recovery due to abandoned parallel
commits.

New metrics:
- `txnrecovery.attempts.pending`
- `txnrecovery.attempts.total`
- `txnrecovery.successes.committed`
- `txnrecovery.successes.aborted`
- `txnrecovery.successes.pending`
- `txnrecovery.failures`

Release note: None

nvb committed:
storage/txnrecovery: add transaction recovery metrics

This commit adds a series of time series metrics that provide insight
into the process of transaction recovery due to abandoned parallel
commits.

New metrics:
- `txnrecovery.attempts.pending`
- `txnrecovery.attempts.total`
- `txnrecovery.successes.committed`
- `txnrecovery.successes.aborted`
- `txnrecovery.successes.pending`
- `txnrecovery.failures`

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37807)
<!-- Reviewable:end -->

nvb committed:
storage: remove needsStaging helper

The function used to contain more logic, but now it just defers to
IsParallelCommit, so there's no point in keeping it.

Release note: None

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37807#-:-LflLohY30oNsKkF9aNk:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/storage/txnrecovery/manager_test.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37807#-LflLSvoEwXoN4LhdTdM:-LflLSvsB4qLJIvhVlTM:b-1u3iil) ([raw file](https://github.com/cockroachdb/cockroach/blob/817a2ab1975b47c39d06a748fd3145ab9f76c66e/pkg/storage/txnrecovery/manager_test.go#L53)):*
> ```Go
> }
> 
> type metricVals struct {
> ```

This pattern is handy but verbose. It’d be nice to automate the generation of these from a metrics struct. Could be a good little Friday project. 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37807#-:-LfptxZ73z_vfqzRuw7Y:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/storage/txnrecovery/manager_test.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37807#-LflLSvoEwXoN4LhdTdM:-Lfpt1zi3eL2o2owDzuf:b-h0y82v) ([raw file](https://github.com/cockroachdb/cockroach/blob/817a2ab1975b47c39d06a748fd3145ab9f76c66e/pkg/storage/txnrecovery/manager_test.go#L53)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This pattern is handy but verbose. It’d be nice to automate the generation of these from a metrics struct. Could be a good little Friday project. 
</blockquote></details>

Yeah, this seems like a cool idea. Code gen might be a bit of a big hammer though. I bet we could do something a little more lightweight with reflection and a `map[string]int64`. Let's keep this in mind for the next time we turn to this pattern.


<!-- Sent from Reviewable.io -->
